### PR TITLE
Fixes to platform's navigation

### PR DESCRIPTION
**Pull request by nramadas**
Two commits, will go in individually.

The first ensures that the handlers will be called even on popstate events. This way, we push the logic of determining whether or not to make api calls to the handlers itself.

The second re-orders the server and shared routes. Server routes are now first, and shared routes come after.

Apologies for the insane diff, I still haven't fixed not checking in compiled code.

:eyeglasses: @schwers @phil303

### CODE REVIEW

File: lib/modules/navigationMiddleware.js [hunk 1]
```diff
@@ -16,44 +16,57 @@ export function parseRoute (path, routes) {
...
-
-          return next(h[method].bind(h));
-        }
-
-        return next(new Error(`No route found for ${method} ${pathName}`));
+        next(action);
+        return next(findAndCallHandler(store, routes, true, action.payload));
+      }
+
+      if (action.type === actions.GOTO_PAGE_INDEX) {
```
Comment (phil303) [inline lib/modules/navigationMiddleware.js]: maybe just use a switch statement here

Comment (phil303) [inline lib/modules/navigationMiddleware.js]: Or maybe something like:

``` js
  let data = action.payload;
  let shouldSetPage = true;
  if (action.type === actions.GOTO_PAGE_INDEX) {
    data = { ...data, method: METHODS.GET };
    shouldSetPage = false;
  }
  next(action);
  return next(findAndCallHandler(store, routes, shouldSetPage, data))
```

**Comment by nramadas** [comment]
:haircut:

File: lib/modules/components.js [hunk 2]
```diff
@@ -364,17 +364,19 @@ export class _UrlSync extends React.Component {
       const currentQuery = extractQuery(self.location.search);
       const currentHash = {}; // TODO: address how hashes are displayed
       let pageIndex = -1;
+      let hist = {};
```
Comment (phil303) [inline lib/modules/components.js:367]: random question. do we normally initialize variables this way? Can't we just declare it?

Comment (nramadas) [inline lib/modules/components.js:367]: i use initialization as a way to denote what type of thing it is.

File: navigationMiddleware.js [hunk 3]
```diff
@@ -174,19 +174,22 @@ return /******/ (function(modules) { // webpackBootstrap
 	    return function (store) {
 	      return function (next) {
 	        return function (action) {
-	          if (action.type === actions.NAVIGATE_TO_URL) {
-	            next(action);
-	            return next(findAndCallHandler(store, routes, true, action.payload));
+	          switch (action.type) {
+	            case actions.NAVIGATE_TO_URL:
+	              {
```
Comment (phil303) [inline navigationMiddleware.js:179]: nit, c style braces?

**Comment by phil303** [comment]
:nail_care:

File: navigationMiddleware.js (hunk 3 above)
Comment (nramadas) [inline navigationMiddleware.js:179]: that's the compiled code :)

Comment (phil303) [inline navigationMiddleware.js:179]: :homer:

**Comment by phil303** [comment]
:fish: seems kosher. I think @schwers should review as well since this is a new-ish section of code for me

File: lib/modules/actions.js [hunk 4]
```diff
@@ -11,9 +11,17 @@ export const setPage = (url, { urlParams={}, queryParams={}, hashParams={}, refe
   payload: { url, urlParams, queryParams, hashParams, referrer },
 });
 
-export const gotoPageIndex = pageIndex => ({
+export const gotoPageIndex = (
+  pageIndex,
+  pathName,
+  {
+    queryParams={},
```
Comment (schwers-zz) [inline lib/modules/actions.js:18]: 🔕 I like that the default values here provide self-documentation

File: lib/modules/navigationMiddleware.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ import pathToRegex from 'path-to-regexp';
 import { METHODS } from './router';
 import * as actions from './actions';
 
-export function parseRoute (path, routes) {
+export function parseRoute(path, routes) {
```
Comment (schwers-zz) [inline lib/modules/navigationMiddleware.js]: Just a reminder to rename this to something like `matchRoute` or `findRoute`

**Comment by schwers-zz** [comment]
Code LGTM, just had a small naming request we talked about in person. 💅

**Comment by nramadas** [comment]
:haircut:

**Comment by schwers-zz** [comment]
🐠  🎉

**Pull request merged**